app/controllers/histories_controller.rb

```diff
@@ -2,15 +2,60 @@
 
 # HistoriesController: for keeping track of user actions over time
 class HistoriesController < ApplicationController
-  before_action :authenticate_user!
+  before_action :authenticate_user!, :check_role, :check_patient
+  before_action :check_history, only: %i[edit delete]
+  rescue_from ActiveRecord::RecordInvalid, with: :handle_validation_error
 
   # Create a new history route; this is used to create comments on subjects.
   def create
-    redirect_to root_url unless current_user.can_create_subject_history?
+    history = History.new(patient_id: @patient.id,
+                          created_by: current_user.email,
+                          comment: params.permit(:comment)[:comment],
+                          history_type: params.permit(:history_type)[:history_type] || History::HISTORY_TYPES[:comment])
+
+    history.original_comment = history if history.history_type == History::HISTORY_TYPES[:comment]
+
+    # Attempt to save and continue; else if failed redirect to index
+    history.save!
+
+    render(json: history) && return
+  end
 
-    History.create!(patient_id: params.permit(:patient_id)[:patient_id],
+  # "Edits" a history comment - a new history comment is created with the updated comment text and a reference to the id of the original
+  def edit
+    History.create!(patient_id: @patient.id,
                     created_by: current_user.email,
                     comment: params.permit(:comment)[:comment],
-                    history_type: params.permit(:type)[:type] || 'Comment')
+                    history_type: History::HISTORY_TYPES[:comment],
+                    original_comment_id: @history.original_comment_id)
+  end
+
+  # "Deletes" a history comment - does not actually remove the comment from the database
+  # but adds a deleted_by and deleted_reason that show the comment was deleted
+  def delete
+    # mark each version of the history as deleted, not just the most recent one
+    @patient.histories
+            .where(original_comment_id: @history.original_comment_id)
+            .update_all({ deleted_by: current_user.email, delete_reason: params.permit(:delete_reason)[:delete_reason] })
+  end
+
+  private
+
+  def check_role
+    redirect_to root_url unless current_user.can_create_subject_history?
```

Comment:
This controller is generally not handling rendering templated views, it is hit via AJAX calls (correct?). In that sense `redirect_to root_url` is probably not the best way to handle the error case. Instead, you should explicitly tell the front end this call was invalid, forbidden, etc.

Try something like:
```
respond_to do |format|
  format.any { head :forbidden }
end
```
or
```
return head :forbidden
```
This should result in the error toast popping up with the relevant error.

Reply:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/0a0fc4efb2a03ddbcdb9ba0ac01bb4adb922e3e9